katsdpcontroller/product_config.py

```diff
@@ -1238,15 +1240,30 @@ async def from_config(cls, config: Mapping[str, Any]) -> 'Configuration':
                     # There are too many possible exceptions from katportalclient to
                     # try to list them all explicitly.
                     raise SensorFailure(f'Could not get component name for {name}: {exc}') from exc
+
+            full_names = set()
+            for name, stream_config in stream_configs.items():
+                stream_cls = STREAM_CLASSES[stream_config['type']]
+                for sensor in stream_cls._class_sensors:
+                    instrument = stream_config['instrument_dev_name']
+                    full_names.add(sensor.full_name(components, name, instrument))
+            # client.sensor_values can take a list of filters - but it then
+            # makes a separate request for each, in series, which is much
+            # slower.
+            regex = '^(' + '|'.join(re.escape(full_name) for full_name in full_names) + ')$'
+            try:
+                samples = await client.sensor_values(regex)
+            except Exception as exc:
+                raise SensorFailure(f'Could not get sensor values: {exc}') from exc
+
             for name, stream_config in stream_configs.items():
                 stream_cls = STREAM_CLASSES[stream_config['type']]
                 for sensor in stream_cls._class_sensors:
                     instrument = stream_config['instrument_dev_name']
                     full_name = sensor.full_name(components, name, instrument)
-                    try:
-                        sample = await client.sensor_value(full_name)
-                    except Exception as exc:
-                        raise SensorFailure(f'Could not get value for {full_name}: {exc}') from exc
+                    if full_name not in samples:
```

Comment:
I can't really see the logic of this second loop. Can't we just check for full_names == samples to check for the case of a missing sensor, and then just loop through samples to check the status of each ?

Reply:
I'm not following - how would we know the values of `name`, `sensor.name` and `sensor.type` in that case?

Reply:
Why do we need them ? We have already iterated through stream_configs to populate full_names, so we have the names that should appear in samples, so checking to see if a requested sensor doesn't exist is easy (i.e. we don't need to loop through stream_configs again). And then only other thing this block does is check the state of each sample...

Reply:
Expand the unmodified lines in the diff - there are unmodified lines further down the block that need them.

Reply:
Fair enough, I yield, although I do think you could still do type_checking and initial assignment without the duplicated code block by just caching the found sensors from the first loop.